### PR TITLE
feat: index referrers on manifest delete

### LINKED ISSUE #347: Add client-side compatibility for deleting manifest with `subject` field

### PR DESCRIPTION
**Pull request by Wwwsylvia**
Resolves #347 
Signed-off-by: Lixia (Sylvia) Lei <lixlei@microsoft.com>

### CODE REVIEW

File: registry/remote/repository.go [hunk 1]
```diff
@@ -928,11 +928,101 @@ func (s *manifestStore) Exists(ctx context.Context, target ocispec.Descriptor) (
...
+// deleteWithIndexing removes the manifest content identified by the descriptor,
+// and indexes referrers for the manifest when needed.
+func (s *manifestStore) deleteWithIndexing(ctx context.Context, target ocispec.Descriptor) error {
+	if target.MediaType == ocispec.MediaTypeArtifactManifest || target.MediaType == ocispec.MediaTypeImageManifest {
+		if state := s.repo.loadReferrersState(); state == referrersStateSupported {
+			// referrers API is available, no client-side indexing needed
+			return s.repo.delete(ctx, target, true)
+		}
+
+		manifestJSON, err := content.FetchAll(ctx, s, target)
```
Comment (shizhMSFT) [inline registry/remote/repository.go:945]: size limit?

File: registry/remote/repository.go [hunk 2]
```diff
@@ -928,11 +928,101 @@ func (s *manifestStore) Exists(ctx context.Context, target ocispec.Descriptor) (
...
 }
 
+// indexReferrersForDelete indexes referrers for image or artifact manifest with
+// the subject field on manifest delete.
+// Reference: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#deleting-manifests
+func (s *manifestStore) indexReferrersForDelete(ctx context.Context, desc ocispec.Descriptor, manifestJSON []byte) error {
+	type artifact struct {
+		Subject *ocispec.Descriptor `json:"subject,omitempty"`
+	}
+	var manifest artifact
```
Comment (shizhMSFT) [inline registry/remote/repository.go:964]: nit:
```suggestion
	var manifest struct {
		Subject *ocispec.Descriptor `json:"subject"`
	}
```

File: registry/remote/repository.go (hunk 1 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go:948]: nice catch!

File: registry/remote/repository.go (hunk 2 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go]: updated.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**